components/constellation/constellation.rs

```diff
@@ -2405,13 +2467,34 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             .map(|frame| frame.id)
             .collect();
         for frame_id in frame_ids {
-            let evicted = match self.frames.get_mut(&frame_id) {
-                Some(frame) => frame.remove_forward_entries(),
+            let (evicted, current_pipeline) = match self.frames.get_mut(&frame_id) {
+                Some(frame) => (frame.remove_forward_entries(), frame.pipeline_id),
                 None => continue,
             };
+            let mut discarded_history_states = vec!();
             for entry in evicted {
                 if let Some(pipeline_id) = entry.pipeline_id {
-                    self.close_pipeline(pipeline_id, DiscardBrowsingContext::No, ExitPipelineMode::Normal);
+                    if pipeline_id != current_pipeline {
+                        self.close_pipeline(pipeline_id, DiscardBrowsingContext::No, ExitPipelineMode::Normal);
+                    } else {
+                        if let Some(state_id) = entry.state_id {
+                            discarded_history_states.push(state_id);
+                        }
+                    }
+                }
+            }
+
+            if !discarded_history_states.is_empty() {
+                let result = match self.pipelines.get(&current_pipeline) {
+                    Some(pipeline) => {
+                        let msg = ConstellationControlMsg::RemoveStateEntries(current_pipeline,
+                            discarded_history_states);
+                        pipeline.event_loop.send(msg)
+                    },
+                    None => return warn!("Pipeline {} closed after removing state entries", current_pipeline),
+                };
+                if let Err(e) = result {
+                    self.handle_send_error(current_pipeline, e);
```

Comment:
This code would be cleaner if we grouped the past by pipeline id, though it means we'd have to keep track of length separately.